### ISSUE TITLE
Issue using custom as os property

### ISSUE THREAD

**Issue by ezoSresyeK** [chyves/chyves#12]
Believe setting os property as custom results in chyves causing a lookup in "/" for device.map and grub.cfg

Installed archlinux with grub-bhyve as loader and os as arch. GPT partitions in guest.

When os is left as arch, it gets a grub prompt. I can insert commands and get grub to boot the guest properly (commands below):

insmod gpt
linux (hd0,gpt1)/vmlinuz-linux root=/dev/sda3
initrd (hd0,gpt1)/initramfs-linux.img
boot

No idea if insmod is needed.

After that I thought why not set the os as custom and see about using those commands in the grub.cfg. No such luck, whenever that's done the following happens.

Preparing to start guest: archguest
Loading guest parameters... done.
Checking if guest is running... not running.
Checking if VMM resources allocated... not allocated.
Checking which type of loader to use... grub-bhyve.
Generating command string for grub-bhyve... 'os' property set as 'custom', user must supply 'device.map' and 'grub.cfg'.
Checking if 'device.map' and 'grub.cfg' files exists at:
Critical error detected. Exiting for following reason:
'/device.map' file does not exist, this file is required to property boot.

command used is chyves archguest start

Looking at the "Checking if 'device.map'..." line it seems that for some reason it never fills in that part as the dataset where the guest is. So looking at "" and appending a "/device.map" makes it look I'm guessing at "/" for the files it wants.

chyves v0.2.0 2016/09/11
chyves branch: 'master'
chyves dataset version: 0005
chyves guest version: 0300
On FreeBSD 11.0-RELEASE-p1
FreeBSD commit: r306420
FreeBSD version revision: 1100122

chyves was installed via ports (or rather my own custom pkg repo built the port and I installed from there).

**Comment by ezoSresyeK** [comment]
To add to this, I believe in lib/chyves-guest-start @ line 384:385, the function is using $_path and said is possibly an undefined variable, whereas @ line 457 it's using ${_GUEST_mountpoint}/device.map for the location of the device.map. However, the function that's immediately above line 384 takes an arg ($1) and assigns it to a $_path. I am wondering if this is the cause of said issue. 

OT (Question):
I'm not a shell scripting guy, so I am unsure if this is possible. If shell script has the ability to have hash tables, why not create a function per OS type that is supported directly in the code and instead of a huge bunch of if's/elif's you get the OS type, then call the function in the hashtable.

For instance (psuedocode):
os_start_func["Arch"] = function() ... end
os_start_func["Debian"] = function() ... end

$os = get_os_from_config()
os_start_func[$os]($args)

Just a thought as it may speed up things vs a bunch of if/elif's since it'd be a lookup, then an exec, vs a see if this is this until finally it is and then do this (which means if one is the least preferred OS (furthest down in the list of if/elif) then the one who uses that will be stuck with a longer start time since it'll need to traverse the entire set of if/elif to get to them).

Hopefully this helps a little with the issue and possibly imparts a useful idea. Thanks.

**Comment by EpiJunkie** [comment]
@ezoSresyeK thanks for opening this bug report. I will be posting a fix in [0.2.3-dev](https://github.com/chyves/chyves/tree/dev) soon and that will trickle into 0.3.0 which will eventually be updated in ports. Sorry, it was an old variable reference (`$_path`) as you pointed out.

When I tested an Arch guest in May, I believe I followed the [installation guide](https://wiki.archlinux.org/index.php/installation_guide) for a MBR installation. I did not try a GPT installation then but tried on Thursday, it was not as straightforward as the MBR installation. You may want to try a EFI graphical installation (`loader=uefi` and `uefi_console_output=vnc`) for the installation, as that should play nicer with GPT disks than EFI booting with GRUB. I suggest graphical because the installation disk did not want to boot via serial in EFI mode. Afterwards, I imagine, the `uefi_console_output` property can be set back to `serial`.

Thanks for the coding suggestion. I am always looking to improve the speed of this tool. As far as I know the original Bourne shell (`/bin/sh`) does not have the capacity of using hash tables or indirect variable references. To me having a dependency free tool is appealing but limiting. Specially, being able to download chyves from GitHub and the bootrom from Peter Grehan's public FreeBSD and running most VMs is pretty neat. Maybe that is too idealistic.

Thank you for your time on posting the bug report, looking into the code for the problem, and also the coding suggestion.

**Comment by ezoSresyeK** [comment]
@EpiJunkie works great in UEFI mode. Outside of the $_path issue what issue did you run into with EFI arch with grub-bhyve. The reason behind GPT is to avoid the 4 actual partition limitation of MBR (tho it's really solveable by adding more disks, this being a vm after all). There's no real reason I want grub-bhyve to work for this, it was just something I tried when I managed to mess up in the EFI mode enough to lock the keyboard out on the virtualization host (think the resolution there was to _NOT_ use vinagre as the vnc client). In the end, the issue would be resolved with the fix in 0.2.3-dev.

Another odd question/OT/suggestion:
I did look and realized hash tables weren't possible in shell scripting (where one is not using an enhanced shell scripting environment). I do commend that it stay completely in sh with very few dependencies (such as bootroms, obviously bhyve, etc needed for running the vm's, not for "enhanced scripting") and find that a good feature.

Another odd proposal instead and this is a bit lengthy and different I believe is as follows:
Create a new global config option that allows a "custom os config directory".
Create a /lib/os_config (choose naming you like, I'm not really too picky about naming and suck at such).
For each OS supported, create a shell script, and what happens is you check the os config for the vm, then check if a file named perhaps $os.sh exists, and then run the script to get the config from there.
Checking in 2 places however, a place for a user to have some custom os scripts that'll return the strings needed for special OS's that the software doesn't include in it's /lib/os_config directory and the /lib/os_config directory.

This allows for a few things, one can easily build a custom OS script to get later merged back here after testing it. Reduces the if/elif's down and prevents the "least preferred OS" gets longest startup time due to list traversal. That's all assuming that such is even feasible or possible. I guess the downside is more files, and more spawned sh processes. I do also believe this could be used to allow a chyves list supported_os feature as well, which would list supported OS's, and then something like chyves list os arch, which would allow the OS config script to return a definition of what types of installation of guest os it supports (mbr?, gpt?, partition layout? etc).

Perhaps the above should be taken out from under this bug and moved towards a feature request instead or proposal, so sorry if I"m cluttering with useless crud here.

**Comment by EpiJunkie** [comment]
> what issue did you run into with EFI arch with `grub-bhyve`.

No issue, however it is more work to get running and I am kinda of lazy/efficient/seeks-the-path-of-least-resistances-when-the-end-results-are-the-same type of person.

> Outside of the $_path issue 

Thankfully this only impacts `grub-bhyve` guests when the '`os`' is set to '`custom`'. (Am not sure if that was clear.)

> The reason behind GPT is to avoid the 4 actual partition limitation of MBR (tho it's really solveable by adding more disks, this being a vm after all).

This was a consideration at one point, so I added some properties and features. A person could have a disk for each 'partition' on the guest.

```
chyves arch create 200m
chyves arch set cpu=2 ram=4g loader=uefi
chyves arch disk disk0 description "/efi"
chyves arch disk add 8g
chyves arch disk disk1 description "/"
chyves arch disk add 100g
chyves arch disk disk2 description "/home"
chyves arch disk add 2g
chyves arch disk disk3 description "swap"

[user@bhost] ~$ chyves info -zd
Guest/Disks     Size   Description
arch            110G   -
  disk0         200M   /boot
  disk1         8G     /
  disk2         100G   /home
  disk3         2G     swap

```

You may even want to set the following property 'on' if you have more a lot disks (see [man page](http://htmlpreview.github.com/?https://raw.githubusercontent.com/chyves/chyves/master/man/chyves.8.html) for the definition):

`chyves global set consolidate_bhyve_pci_devices=yes`

> vnc client

I have had good success across multiple platforms using [TightVNC](http://www.tightvnc.com/) Java client.

#### OT:

>  I do commend that it stay completely in sh with very few dependencies ... and find that a good feature.

Thank you.

> I guess the downside is more files, and more spawned sh processes.

That was my thought as well. While I have not tested it, I can not imagine the `sh` spawns would save time over `if`/`elif`s. It would also create another set of checks that would be needed to be ran to see if the files exist. Neat ideas, something I had not considered.

> This allows for a few things, one can easily build a custom OS script to get later merged back here after testing it.

My hope is that it being a shell script, it would be simple enough for most people to add an `elif` statement and send patches or PRs.

I could create a new '`os`' type called '`command`' that could run a script located in the guest directory. This might be the simplest solution. Thoughts?

>  I do also believe this could be used to allow a `chyves list supported_os` feature as well, which would list supported OS's, and then something like `chyves list os arch`, which would allow the OS config script to return a definition of what types of installation of guest os it supports (mbr?, gpt?, partition layout? etc).

Definitely an interesting idea. Not sure I would like managing two points with that information ([man page](http://htmlpreview.github.com/?https://raw.githubusercontent.com/chyves/chyves/master/man/chyves.8.html) and a `list` sub-command). That said, I have considered adding a matrix of tested OSes and ISOs for the `grub-bhyve` guests. I do have mixed feeling however, most of them revolving around the simplicity that UEFI, UEFI-GOP, and UEFI-CSM add to booting guests compared to `grub-bhyve` guests. I think the easiest place to add that information would be on [chyves.org](http://chyves.org/).

> Perhaps the above should be taken out from under this bug and moved towards a feature request instead or proposal, so sorry if I"m cluttering with useless crud here.

No worries, I _really_ appreciate the feedback and ideas.

**Comment by ezoSresyeK** [comment]
> ```
> Outside of the $_path issue
> ```
> 
> Thankfully this only impacts grub-bhyve guests when the 'os' is set to
> 'custom'. (Am not sure if that was clear.)

It was clear, I ended up setting the os back to arch and then in console mode
entering in the grub commands instead.

> This was a consideration at one point, so I added some properties and features.
> A person could have a disk for each 'partition' on the guest.
> 
> chyves arch create 200m
> chyves arch set cpu=2 ram=4g loader=uefi
> chyves arch disk disk0 description "/efi"
> chyves arch disk add 8g
> chyves arch disk disk1 description "/"
> chyves arch disk add 100g
> chyves arch disk disk2 description "/home"
> chyves arch disk add 2g
> chyves arch disk disk3 description "swap"
> 
> [user@bhost] ~$ chyves info -zd
> Guest/Disks     Size   Description
> arch            110G   -
>   disk0         200M   /boot
>   disk1         8G     /
>   disk2         100G   /home
>   disk3         2G     swap
> 
> You may even want to set the following property 'on' if you have more a lot
> disks (see man page for the definition):
> 
> chyves global set consolidate_bhyve_pci_devices=yes

Fits in with the lazy/efficient/etc approach though could be viewed as sort of
hackish, though it definitely does work. Not meant as multi-disk support is
hackish but more that using it instead of GPT partitions.

> ```
> vnc client
> ```
> 
> I have had good success across multiple platforms using TightVNC Java client.

I installed tightvnc from ports and I guess that's not the java viewer but the
issue I had with it is the display would black out at certain points, after say
using a VT under FreeBSD if you then woke the display up it'd still leave a lot
of things blacked out except like where the cursor was currently. Currently
using KRDC and wouldn't mind finding something that worked utilizing less
resources. I've been wondering on integration with a web capable vnc client as
well.

> OT:
> 
> ```
> I do commend that it stay completely in sh with very few dependencies ...
> and find that a good feature.
> ```
> 
> Thank you.
> 
> ```
> I guess the downside is more files, and more spawned sh processes.
> ```
> 
> That was my thought as well. While I have not tested it, I can not imagine the
> sh spawns would save time over if/elifs. It would also create another set of
> checks that would be needed to be ran to see if the files exist. Neat ideas,
> something I had not considered.

The key moreso to that I'd guess is if the list of if/elif grew long enough
you're looking at linear time to start the last supported OS whereas with the
script approach per OS you're looking at "constant" time barring considerations
of other variables such as CPU utilization, disk load, string comparison
operators, etc. Can't say it's at that point and later in here I'll address why
I believe it'll never approach that point actually which means this proposal
though sensible "possibly" would be a lot of void work for little benefit after
further consideration.

> ```
> This allows for a few things, one can easily build a custom OS script to
> get later merged back here after testing it.
> ```
> 
> My hope is that it being a shell script, it would be simple enough for most
> people to add an elif statement and send patches or PRs.

The issue there is one has to know what variables you use and how, whereas if
you simplify it where you call a script and they return a string or strings they
don't have to know all that deeply how chyves works. However, the work for such
might not be worth the effort as addressed a little further down.

> I could create a new 'os' type called 'command' that could run a script located
> in the guest directory. This might be the simplest solution. Thoughts?

While that would work, it's almost like one creating their own custom device.map
or grub.cfg (which doesn't work when os="custom"). So it'd kind of be like
redoing the custom OS feature. If the $_path gets changed around to work
properly again then the command option might not be needed. I can't think of a
case where it would be needed when one can supply their own version of those 2
files atm at least (haven't thought too deeply, and definitely not knocking it
as a feature because it does have the option where it could integrate with guest
helped deployments for automation purposes for like if someone was hosting vm's
for customers via bhyve possibly).

> ```
> I do also believe this could be used to allow a chyves list supported_os
> feature as well, which would list supported OS's, and then something like
> chyves list os arch, which would allow the OS config script to return a
> definition of what types of installation of guest os it supports (mbr?,
> gpt?, partition layout? etc).
> ```
> 
> Definitely an interesting idea. Not sure I would like managing two points with
> that information (man page and a list sub-command). That said, I have
> considered adding a matrix of tested OSes and ISOs for the grub-bhyve guests. I
> do have mixed feeling however, most of them revolving around the simplicity
> that UEFI, UEFI-GOP, and UEFI-CSM add to booting guests compared to grub-bhyve
> guests. I think the easiest place to add that information would be on
> chyves.org.

The key to the list command displaying supported OS's was that it'd just call
the OS script in like /lib/os_config with the argument "define" and it'd return
the definition of support, so it wouldn't really be I think a case where the man
page would need updating because if one supplied their own os configs then the
man page wouldn't match and it would depend on if builtin os configs or user os
configs take precedence. Testing and listing ISO's for grub-bhyve would be a
continual catch up game and quite a large amount of testing.

However, you sorta hit the nail on the head with your mixed feelings. UEFI,
UEFI-GOP, and possibly UEFI-CSM (which I couldn't get to work) are where I
believe the future is headed with bhyve truthfully. That allows true user
managed OS installations vs host managed or compliant to a specific way of doing
things. The only reason I ever went for grub-bhyve is because I managed to lock
out my keyboard due to UEFI-GOP + UEFI + VNC. The only issue with all of that
also is currently UEFI-GOP consumes CPU quite a bit _when an vnc client is
connected_, whereas if I run a vnc server within a guest the cpu useage is
negligible and relegated to the guest (which could be a customer). I have not
tried UEFI console nor got UEFI-CSM to work however. I do think instead of
supporting grub-bhyve it might make better sense to relegate it to custom only
and recommend UEFI utilization... of course if I find another keyboard lockup
or bhyve issue my opinion could flop till said is fixed. With UEFI being the
most supported and recommended path and the fact it doesn't need custom files
each and every time possibly for a custom OS it would also satisify my thoughts
on near constant vm launch times. Perhaps there is a grand reason to continue to
support heavilly and even recommend grub-bhyve that I'm overlooking being that I
just started playing with bhyve a couple of weeks ago due to UEFI-GOP and
FreeBSD 11 (didn't wish to attempt under FreeBSD 10). The other thing to
consider I guess is UEFI-GOP FB Size being at certain sizes will sometimes cause
a boot hangup it seems for no specific reason (FreeBSD11 ISO works with
1920x1080, installed base wouldn't work with same resolution but worked with a
higher resolution however).
